### PR TITLE
DependencyContext step 2: Add compilation options

### PR DESCRIPTION
**Pull request by pakrym**
@anurse  @davidfowl

### CODE REVIEW

File: src/Microsoft.Extensions.DependencyModel/DependencyContextWriter.cs [hunk 1]
```diff
@@ -22,10 +23,56 @@ public void Write(DependencyContext context, Stream stream)
...
                 new JProperty(DependencyContextStrings.LibrariesPropertyName, WriteLibraries(context))
                 );
+        }
 
+        private JObject WriteCompilationOptions(CompilationOptions compilationOptions)
+        {
+            var o = new JObject();
+            if (compilationOptions.Defines != null)
+            {
+                o.Add(new JProperty(DependencyContextStrings.DefinesPropertyName, compilationOptions.Defines));
```
Comment (davidfowl) [inline src/Microsoft.Extensions.DependencyModel/DependencyContextWriter.cs]: `o[DependencyContextStrings.DefinesPropertyName] = compilationOptions.Defines;`

File: src/Microsoft.Extensions.DependencyModel/DependencyContext.cs [hunk 2]
```diff
@@ -24,6 +25,8 @@ public DependencyContext(string target, string runtime, Library[] compileLibrari
 
         public string Runtime { get; set; }
```
Comment (davidfowl) [inline src/Microsoft.Extensions.DependencyModel/DependencyContext.cs]: Runtime and Library should be get only

File: src/Microsoft.Extensions.DependencyModel/CompilationOptions.cs [hunk 3]
```diff
@@ -0,0 +1,50 @@
+﻿using System.Collections.Generic;
+
+namespace Microsoft.Extensions.DependencyModel
+{
+    public class CompilationOptions
```
Comment (davidfowl) [inline src/Microsoft.Extensions.DependencyModel/CompilationOptions.cs]: It's unfortunate that we need to duplicate this model.... Can we source share?

File: src/Microsoft.DotNet.ProjectModel/DependencyContextBuilder.cs [hunk 4]
```diff
@@ -5,38 +5,72 @@
...
 
-        private static Library GetLibrary(LibraryExport export, IEnumerable<LibraryAsset> libraryAssets, IEnumerable<LibraryExport> dependencies)
+        private static Library[] GetLibraries(IEnumerable<LibraryExport> dependencies, NuGetFramework target, string configuration, Func<LibraryExport, IEnumerable<LibraryAsset>> assemblySelector)
         {
+            return dependencies.Select(export => GetLibrary(export, target, configuration, assemblySelector(export), dependencies)).ToArray();
+        }
+
+        private static Library GetLibrary(LibraryExport export, NuGetFramework target, string configuration, IEnumerable<LibraryAsset> libraryAssets, IEnumerable<LibraryExport> dependencies)
+        {
+            var type = export.Library.Identity.Type.ToString().ToLowerInvariant();
```
Comment (davidfowl) [inline src/Microsoft.DotNet.ProjectModel/DependencyContextBuilder.cs]: why do you ToString() type?

File: src/Microsoft.Extensions.DependencyModel/CompilationOptions.cs (hunk 3 above)
Comment (pakrym) [inline src/Microsoft.Extensions.DependencyModel/CompilationOptions.cs]: Other model is mutable and has more properties that doesn't make sense (preserver compilation context)

**Comment by pakrym** [comment]
:up: :date:

File: src/Microsoft.DotNet.ProjectModel/DependencyContextBuilder.cs [hunk 5]
```diff
@@ -5,38 +5,72 @@
...
+                compilerOptions.KeyFile,
+                compilerOptions.DelaySign,
+                compilerOptions.PublicSign,
+                compilerOptions.EmitEntryPoint);
         }
 
-        private static Library GetLibrary(LibraryExport export, IEnumerable<LibraryAsset> libraryAssets, IEnumerable<LibraryExport> dependencies)
+        private static Library[] GetLibraries(IEnumerable<LibraryExport> dependencies, NuGetFramework target, string configuration, Func<LibraryExport, IEnumerable<LibraryAsset>> assemblySelector)
         {
+            return dependencies.Select(export => GetLibrary(export, target, configuration, assemblySelector(export), dependencies)).ToArray();
```
Comment (davidfowl) [inline src/Microsoft.DotNet.ProjectModel/DependencyContextBuilder.cs]: This looks really inefficient. Couldn't this be done with a single foreach loop over all dependencies?

**Comment by piotrpMSFT** [comment]
@dotnet-bot test this please

File: src/Microsoft.DotNet.ProjectModel/DependencyContextBuilder.cs [hunk 6]
```diff
@@ -5,49 +5,95 @@
...
+        private static Dependency GetDependency(LibraryRange libraryRange, IDictionary<string, Dependency> dependencyLookup)
         {
-            var version =
-                dependencies.First(d => d.Library.Identity.Name == libraryRange.Name)
-                    .Library.Identity.Version.ToString();
-            return new Dependency(libraryRange.Name, version);
+            Dependency dependency;
+            if (!dependencyLookup.TryGetValue(libraryRange.Name, out dependency))
+            {
+                throw new InvalidOperationException($"Can not find dependency info for {libraryRange.Name}, try runnung \"dotnet restore\"");
```
Comment (davidfowl) [inline src/Microsoft.DotNet.ProjectModel/DependencyContextBuilder.cs]: Remove this

**Comment by davidfowl** [comment]
@dotnet-bot test this please

**Comment by davidfowl** [comment]
:shipit: Once the build passes

**Pull request merged**